### ISSUE TITLE
Provide an endpoint for character's associated fleet_id

### ISSUE THREAD

**Issue by xSetech** [ccpgames/esi-issues#155]
Augment an existing or provide a new endpoint which yields the id of the fleet a character has joined.

Enabling the development of fleet management and communication utilities are obviously a goal of ESI. There are many operations provided for fleet manipulation. These all rely on a provided fleet_id. The easier and more automated the acquisition of that id, the better the derived utilities.

A cursory search does not appear to explain why this is not implemented in the traditional API, CREST, or ESI currently.

**Comment by Regner** [comment]
> A cursory search does not appear to explain why this is not implemented in the traditional API, CREST, or ESI currently.

It was not implemented in other APIs because it is not stored in the DB, so the XML API cannot retrieve it, and getting to it within the cluster would have required more hops throughout the cluster and putting some load on the location/fleet nodes. For it to be useful it would need to be on a really low cache time and would likely have resulted in a large number of requests (everyone wanting to track if characters enter a fleet) so was rejected at the time. On top of that as it currently stands a fleet boss just needs to take one action and the tracking of everyone in that fleet can happen at once without a huge number of extra requests so it wasn't really deemed as a needed feature.

**Comment by Sciencegeek123** [comment]
@Regner would it be more feasible to get delayed fleet history for a player? Perhaps compiled at downtime? That should address the latency and load concerns. This would allow automation of SRP and PAPs, as well as provide useful corp activity statistics.

**Comment by xSetech** [comment]
@Regner Thank you for your quick and almost thoroughgoing response. I understand unoptimized indices are a very strong barrier to overcome without extensive engineering effort. Since that is the case, this Issue should either pivot to a discussion of related problems and usefulness or just be closed. The feature may not be the most appropriate solution to help automate fleet and player management as @Sciencegeek123 points out (thank you for your input as well).

Fleets are critical in a game about spaceships, being logical groups of ship activity. If players, fleet bosses, and users of the fleet interface have a quick way to gauge other player "activity", based on whether a player is in a fleet or not, the pre-fleet procedures would be improved- particularly as corporations scale. In that case, the feature is reduced down to providing whether or not a player is in a fleet at all. From there, it can be gleaned that a particular player is ready to be invited to a newly created fleet op, and leaders have some method of judging engagement in whatever operations they host. In my view, an essential metric.

In conclusion, while this feature request is clearly not a barrier to making use of the existing fleet endpoints provided by ESI, it may be a hint to another feature that is easier to implement and just as useful.

**Comment by ErikKalkoken** [comment]
I like to support the request of providing access to the fleet_ID via API endpoint. Currently the FC has to copy&paste a URL from the client into an external tool in order to use it. This manual step will be seen as inconvenient by many FCs and most likely reduce acceptance and frequent usage of fleet tools.

To address concerns about load I would suggest to make that information accessible to authed users users only. Only the fleet FC need to be able to access it.

e.g.
`/characters/[character_id}/fleet`
with a new required score: `esi-fleets.read_character_fleet.v1`

**Comment by a-tal** [comment]
game design has signed off on this request.

if you're in a spy fleet, or want to otherwise keep your fleet members private from someone with a read_fleet scope, don't be in a squad, wing, or fleet command position (or the fleet boss) which exposes the members, locations, and ship types.

what will be exposed from your spy fleet are: the motd, if free move is on, if the fleet is advertised, if voice comms are enabled, and the wing and squad names.

**Comment by CarbonAlabel** [comment]
The request here was just for an endpoint to get your current fleet ID, you're talking about making it possible to get (parts of) fleet info for people other than the fleet boss. Just to be sure, we're getting both?

**Comment by DaneelTrevize** [comment]
Can we get clarification of any feature creep mentioned here? This was originally just to remove the need for a player to have to manually Export Fleet Link URL and paste it into API apps.  
A reasonable improvement would simple be a toggle option in-game instead, to expose the FleetID under the character via the API, so a boss can choose if a given fleet is exposed or remains private.  
No other changes were needed, no extra exposure of fleet details for non-boss characters via API.

**Comment by a-tal** [comment]
feature creep? I'm just talking about what you could do with the rest of the existing API given a character_id->fleet_id route. what you're describing with some in-game toggle affecting ESI would be feature creep... if you're passionate about that feature, please open a new issue for it, with some use case for it.

this is in staging atm btw, found a bug with the action processor last week that will need a monolith patch before we can push this to production. the model for the return is:

```json
{
  "fleet_id": 1234567890,
  "wing_id": -1,
  "squad_id": -1,
  "role": "fleet_commander",
}
```

`role` is an enum of `["fleet_commander", "wing_commander", "squad_commander", "squad_member"]`, everything else is an `int64`.

**Comment by DaneelTrevize** [comment]
Perhaps you can help answer some of these questions.
Since when has a non-fleet-boss been able to generate the External Fleet Link?
Since when has a non-fleet-boss been able to access fleet details via the API, regardless of how they obtain the FleetID?
Where is the documentation of how expected return codes changed from 403 to 200 given this fleet command position condition?
Where is the documentation of the extent of the esi-fleets.read_fleet.v1 scope?
And if it's changed, why hasn't the scope version increased?

**Comment by DaneelTrevize** [comment]
OK, I've just tested this. As a fleet boss you can currently find the External Fleet Link menu entry, and can use the API with the read scope authorised. Done via the CCP hosted interface.
Then after you make someone else boss of that same fleet, you get 404 "error": "The fleet does not exist or you don't have access to it!". Regardless of which member/squad/wing/fleet command position, non-bosses don't see a menu entry. With the boss restored to myself, the API again works for the same exact FleetID.

So currently the scope does not give access to extra fleet details for non-boss, and command positions grant the in-game Fleet Composition menu entry but not the Generate External Link, and don't permit API access of related data.

So talking about all non-members gaining the FleetID under their character API, and that then granting rights, would be a change & one that's not being asked for in this issue. Thus I call it scope creep.

All we want is for the boss to not have to click the External Link and then manually paste that into an app.
I understand that with no additional per-fleet-instance control this would enable an app with granted scope & token to poll for every fleet the character becomes boss of. But that's far more limited than being discussed by CCP devs here, and also why I propose a toggle option to let the player choose per fleet instance whether the fleet appears/is accessible via API.

**Comment by ddouglas493** [comment]
Alright @DaneelTrevize. I just read, and reread what @a-tal posted above. With no experience with the fleet endpoints whatsoever I was able to grasp what he meant, so I will spell it out below, but before I do I must say you are being very disrespectful and idiotic to a member of CCP that has been one of the most responsive members of CCP to the Development community for the better part of a year, not leaving out the rest of the Team Tech Co, but I see more of Snoewden than the others. Sorry.

So here we go:

He is saying that if the Fleet Boss has authenticated with an application that has the read_fleet scope, the application can now retrieve the fleet ID without input from the Fleet Boss and then query details about the fleet from the Fleet Endpoint. This is EXACTLY what this thread is asking for. To farther expand on his post, he said that if the Fleet Boss has authenticated with the same application that has the Fleet Scope and he has a fellow corp mate, friend, whatever in fleet with him that he KNOWS is a spy, than that player DOES NOT need to be in a Command Position otherwise the application will be able to interact with that character in some form or manner. Like I said, I don't have experience with the Fleet Endpoint, so I do not know what it is capable of, but I believe this is what @a-tal means. If I am wrong, than that is 5 minutes I am never going to get back. 

Fly Safe and stop being stupid. Get on Slack as well, don't do shit like this on GH Issues.

**Comment by DaneelTrevize** [comment]
"that is 5 minutes I am never going to get back"

@a-tal makes the first mention of whatever a "spy fleet" is, and describes non-boss fleet positions exposing basically what the Fleet Composition windows shows to non-members iirc. This is obviously visible to those positions in-game (tautology), so my understanding is he's claiming that read scope will enable such data via the API. Yet I retested and it currently does not.
You can parrot the devs and claim it makes sense, yet I'll just repeat the evidence that it does not.

**Comment by ddouglas493** [comment]
![](https://i.imgur.com/ZM4ctJK.png)

**Comment by Aidansavage** [comment]
Regarding your "testing," Daneel, you should have actually read what a-tal posted instead of being some snide twit trying to one up a dev who is actually /trying/ to do stuff for eve. If you had, you'd have realized, and noticed for once, that he said it was "in staging." That's distinctly different from "it's live." It's also an extremely important one that anyone who claims to be competent with coding should know.

**Comment by DaneelTrevize** [comment]
@ddavaham Not sure why you're still addressing just me, this pic seems to confirm a-tel/SnowedIn contradicting himself as per https://github.com/ccpgames/esi-issues/issues/155#issuecomment-327271063 and confirming that the API indeed functions the way I've stated, which is good news for all.

@Aidansavage GG white-knighting, but we've just had evidence that the devs now agree the API _won't_ do (i.e. doesn't do in staging) what I'm saying it already _doesn't_ do live, so any statements about "spies" not being in certain fleet command positions is _completely irrelevant to this API discussion, as those positions only affect intel access via in-game features_. Thankfully this means we're back to not adding more unspecified divergence between in-game just-fleet-member intel access and API intel access.

Again, surely any proposal w.r.t. not-boss, not-just-fleet-member characters having more read access via API would need to address the point that in-game currently all fleet members can see the list/hierarchy of all fleet members & squad/wing names, which AFAIK isn't being proposed via the API (even if something similar to the Fleet Composition feature for applicable members inadvertently is).

**Comment by a-tal** [comment]
This thread is way off the rails. There will be no attacking of members here. If you have a gripe with someone take it out in New Eden.

Stop putting words in my mouth, then claiming when I clarify (read: saying the same thing with different words) my original message that is somehow confirmation of contradicting something.

If you have some general grievance with EVE, or ESI, or me, or whatever, then take that up in the forums, on reddit, on slack, whatever. Not here where ~100 developers are notified.

I'm locking this thread. Will post updates in it for the release of the model outlined above. And will, as always, be available on slack to discuss any details related to this or any other route in ESI.
